Add docs and additional tests

sanjay990 opened the pull request:
This commit adds docs and additional java based tests for SQL Server.

akshatnair commented:
Looks good, most of it seems pretty similar to MySQL.

sanjay990 commented:
@akshatnair : Yep, its almost same, though we don't have negative and data type tests for other JDBC based connectors.

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+
+The SQL Server connector can only access a single database within
+a SQL Server server. Thus, if you have multiple SQL Server databases,
+or want to connect to multiple SQL Server servers, you must configure
```
_want to connect to multiple instance of the SQL Server_

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+The SQL Server connector can only access a single database within
+a SQL Server server. Thus, if you have multiple SQL Server databases,
+or want to connect to multiple SQL Server servers, you must configure
+multiple instances of the SQL Server connector.
```
`catalogs` instead of `instances of the connector`?

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+
+To add another catalog, simply add another properties file to ``etc/catalog``
+with a different name (making sure it ends in ``.properties``). For example,
+if you name the property file ``sales.properties``, Presto will create a
```
It's nothing SQL server specific. Shouldn't we refer to the general Presto docs here?

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+Querying SQL Server
+-------------------
+
+The SQL Server connector provides a schema for every SQL Server schema.
```
Maybe something like: "_The SQL Server connector provides access to all schemas visible to the specified user in configured database._"?

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+The SQL Server connector provides a schema for every SQL Server schema.
+You can see the available SQL Server schemas by running ``SHOW SCHEMAS``::
+
+    SHOW SCHEMAS FROM sqlserver;
```
Again, isn't it too generic?

maciejgrzybek replied:
Also, you don't refer name `sqlserver`. You should say e.g. "for SQL Server catalog named _sqlserver_ this would be:"

sanjay990 replied:
I think its fine to have it here as an example. In case, you want I can provide a link to `show schemas`

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+
+    SHOW SCHEMAS FROM sqlserver;
+
+If you have a SQL Server schema named ``web``, you can view the tables
```
_SQL Server_ is unnecessary.

brian-rickman commented:
LGTM

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+    DESCRIBE sqlserver.web.clicks;
+    SHOW COLUMNS FROM sqlserver.web.clicks;
+
+Finally, you can access the ``clicks`` table in the ``web`` schema::
```
Why you recognize _access_ to be _select_ only?

sanjay990 replied:
This is a generic which we use for all JDBC based connectors.
What else do you plan to include.

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+
+    SELECT * FROM sqlserver.web.clicks;
+
+If you used a different name for your catalog properties file, use
```
`sqlserver` catalog name should be mentioned before.

sanjay990 replied:
didn't get this

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+smallint
+int
+tinyint
+float
```
Shouldn't we mention the difference / lack of difference between `float` and `real`?

sanjay990 replied:
Rather than explaining the difference between different SQL Server data types. I can provide a link for the same.

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -0,0 +1,98 @@
+real
+char(n)
+varchar(n)
+text
```
What type we translate `text` into? I believe we should have this explained with any details whenever they're relevant. E.g. I'd picture this as a table:
SQL Server type | Presto type
float double
real float

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/NegativeTests.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Test(groups = SQL_SERVER)
+    public void testInvalidTable()
```
s/Invalid/NonExistent/

sanjay990 replied:
57e4fb11d7877a38d0f5852c125a21e2db2dd2cb

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/NegativeTests.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Test(groups = SQL_SERVER)
+    public void testInvalidSchema()
```
Ditto

sanjay990 replied:
57e4fb11d7877a38d0f5852c125a21e2db2dd2cb

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/NegativeTests.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Test(groups = SQL_SERVER)
+    public void testInvalidColumn()
```
Ditto

sanjay990 replied:
57e4fb11d7877a38d0f5852c125a21e2db2dd2cb

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -0,0 +1,102 @@
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return compose(
+                immutableTable(NATION)
```
Unnecessary newlines.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -0,0 +1,102 @@
+    {
+        String sql = format(
+                "SELECT n_nationkey, n_name, n_regionkey, n_comment FROM %s.%s.%s",
+                CONNECTOR_NAME,
```
Extract `format("%s.%s.%s", CONNECTOR_NAME, KEY_SPACE, NATION.getName())` to constant and reuse it in all tests here instead of copy&pasting.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test(groups = SQL_SERVER)
+    public void testNationJoinNation()
```
`testNationSelfInnerJoin`

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -0,0 +1,102 @@
+    @Override
+    public Requirement getRequirements(Configuration configuration)
+    {
+        return compose(
```
Do we need a `compose` here?
You didn't use it here: https://github.com/Teradata/presto/pull/503/commits/01090b1f873a3ff73ba983e5e86e0d9d1cc887cc#diff-96992044a8263d5b64ebe0e853fcb3b3R40

sanjay990 replied:
Yes, its needed if we need to create multiple tables.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -0,0 +1,102 @@
+
+        assertThat(queryResult).containsOnly(row("CANADA", "AMERICA"));
+    }
+}
```
Add a test for pushdown (`where` clause).
Also for aggregates (e.g. `count(1)`) because we've seen issues with them previously in other connectors.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -42,7 +55,8 @@
-                immutableTable(NATION)
+                immutableTable(NATION),
+                immutableTable(SQLSERVER_ALL_TYPES)
         );
```
Wrong formatting.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -99,4 +113,30 @@ public void testNationJoinRegion()
+                        row(Long.MIN_VALUE, Short.MIN_VALUE, Integer.MIN_VALUE, Byte.MIN_VALUE, Double.MIN_VALUE,
+                                Float.MIN_VALUE, "\0   ", "\0", "\0", "\0   ", "\0", "\0",
+                                Date.valueOf("0001-01-02"), Timestamp.valueOf("1753-01-01 00:00:00.000"),
+                                Timestamp.valueOf("0001-01-01 00:00:00.000"), Timestamp.valueOf("1900-01-01 00:00:00")),
```
Is it properly defined by JDBC standard what's the expected value for such a date?

sanjay990 replied:
This is supported by respective SQL Server data type. For e.g. for date
`0001-01-01 through 9999-12-31 `
https://msdn.microsoft.com/en-us/library/bb630352.aspx

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -99,4 +113,30 @@ public void testNationJoinRegion()
+                                Float.MIN_VALUE, "\0   ", "\0", "\0", "\0   ", "\0", "\0",
+                                Date.valueOf("0001-01-02"), Timestamp.valueOf("1753-01-01 00:00:00.000"),
+                                Timestamp.valueOf("0001-01-01 00:00:00.000"), Timestamp.valueOf("1900-01-01 00:00:00")),
+                        row(Long.MAX_VALUE, Short.MAX_VALUE, Integer.MAX_VALUE, Byte.MAX_VALUE, Double.MAX_VALUE,
```
How about the test of parametrized `float`?

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/SqlServerDataTypesTableDefinition.java:
```diff
@@ -0,0 +1,62 @@
+
+    private static final String ALL_TYPES_DDL =
+            "CREATE TABLE %NAME% (bi bigint, si smallint, i int, ti tinyint, f float, r real," +
+                    "c char(4), vc varchar(4), te text, nc nchar(4), nvc nvarchar(4), nt text," +
```
Why all of precision 4 only?

sanjay990 replied:
I wanted to test with some arbitrary, just took 4.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/SqlServerDataTypesTableDefinition.java:
```diff
@@ -0,0 +1,62 @@
+                            Timestamp.valueOf("0001-01-01 00:00:00.000"), Timestamp.valueOf("1900-01-01 00:00:00")),
+                    ImmutableList.of(Long.MAX_VALUE, Short.MAX_VALUE, Integer.MAX_VALUE, Byte.MAX_VALUE,
+                            Double.MAX_VALUE, Float.MAX_VALUE, "abcd", "abcd", "abcd", "abcd", "abcd", "abcd",
+                            Date.valueOf("9999-12-31"), Timestamp.valueOf("9999-12-31 23:59:59.997"),
```
Are you sure you want to rely on the `Timestamp.valueOf()` for non-standard timestamps?
Is the behavior of `Timestamp.valueOf()` in such a case what we want to expect from that connector?

maciejgrzybek commented on presto-product-tests/README.md:
````diff
@@ -244,6 +244,9 @@ be any one of the available profiles:
     presto-product-tests/bin/run_on_docker.sh <profile> -x quarantine,big_query,profile_specific_tests
     ```
 
+Note: SQL Server product-tests use `microsoft/mssql-server-linux` docker container. We assume you accept the license
````
I don't know how it should sound from the legal perspective but "we assume" does not sound good to me. Maybe "By running sql server product tests you accept the license ..." is better? Or someone with actual legal knowledge should take a stand here?

sanjay990 commented:
@maciejgrzybek : Incorporated your comments. 
I will merge data type and java tests commits after your LGTM :)

I added a new commit 9cdd6ea16e70bef6fbfb10e3e759d39e0a922787 to remove the catalog creation from MySql and PostgreSQL

maciejgrzybek commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -28,23 +28,22 @@ Multiple SQL Server Databases or Servers
 a SQL Server server. Thus, if you have multiple SQL Server databases,
-or want to connect to multiple SQL Server servers, you must configure
-multiple instances of the SQL Server connector.
+or want to connect to multiple instance of the SQL Server, you must configure
```
`instances`

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -60,32 +60,34 @@ public Requirement getRequirements(Configuration configuration)
-                NATION.getName());
+                "SELECT n_nationkey, n_name, n_regionkey, n_comment FROM %s",
+                NATION_TABLE_NAME
+        );
```
Wrong formatting (put ")" in the end of previous line).

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -60,32 +60,34 @@ public Requirement getRequirements(Configuration configuration)
-                tableName);
+                NATION_TABLE_NAME,
+                NATION_TABLE_NAME
+        );
```
ditto formatting

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -102,12 +104,11 @@ public void testNationJoinRegion()
-                KEY_SPACE,
-                NATION.getName());
+                NATION_TABLE_NAME
+        );
```
ditto

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java:
```diff
@@ -41,7 +41,7 @@ public Requirement getRequirements(Configuration configuration)
     }
 
     @Test(groups = CASSANDRA)
```
Why did you change cassandra tests while it's sql server PR?

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/sqlserver/Select.java:
```diff
@@ -120,24 +119,28 @@ public void testAllDatatypes()
-                                Timestamp.valueOf("9999-12-31 23:59:59.999"), Timestamp.valueOf("2079-06-06 00:00:00")),
-                        row(null, null, null, null, null, null, null, null, null, null, null, null)
+                                Timestamp.valueOf("9999-12-31 23:59:59.999"), Timestamp.valueOf("2079-06-06 00:00:00"),
+                                Double.valueOf("1234567891234567"), Float.valueOf("12345678")
```
Why don't you use any fraction numbers?
Integers up to 2^24 can be precisely expressed by float (and up to 2^53 for doubles) IIRC.

maciejgrzybek commented on presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerClient.java:
```diff
@@ -42,7 +42,7 @@ public void commitCreateTable(JdbcOutputTableHandle handle)
         StringBuilder sql = new StringBuilder()
                 .append("sp_rename ")
-                .append(singleQuote(handle.getCatalogName(), handle.getSchemaName(), handle.getTableName()))
+                .append(singleQuote(handle.getCatalogName(), handle.getSchemaName(), handle.getTemporaryTableName()))
```
Why? How is it related?

sanjay990 replied:
Its for CTAS, this was a bug.

maciejgrzybek commented on presto-product-tests/src/main/java/com/facebook/presto/tests/cassandra/NegativeTests.java:
```diff
@@ -41,7 +41,7 @@ public Requirement getRequirements(Configuration configuration)
 
     @Test(groups = CASSANDRA)
-    public void testNonExistentTable()
+    public void testInvalidTable()
```
You could simply revert your previous fixup ;)
